DA-289: Refactor and rework handling of line breaks

stgrue opened the pull request:
In order to fix several bugs and undesirable effects, this patch reworks large
parts of how line breaks are handled in the annotation view. Most importantly,
a new system for calculating the width of lines was added, so that the positions
of line breaks can now be determined with greater accuracy instead of relying on
coarse approximations of the average character width. Alongside the more general
refactoring, this has the following effects:
- The application does not crash anymore when the length of a word exceeds the
  length of the annotation box.
- Lines containing many characters of above-average length (e.g. in Chinese texts)
  are not cut off anymore.
- Line breaks are (in general) closer to the edge of the annotation box, resulting
  in a more economic use of horizontal space.
- The new system can be manually optimized for other languages and scripts as well.

judembo commented:
I noticed a few things:

1. Documents with words that are too long for one line are now loaded fine, but the respective words/llines are simply cut off and scrolling to the side is not possible. I don't know an elegant solution for this, but would it be possible to force a line break there and just split the word at the end of the annotation box?

2. When using character wise tokenization, lines are still cut off as can be seen in this screenshot:
![screenshot from 2016-10-12 15 04 49](https://cloud.githubusercontent.com/assets/19438668/19311143/5790137c-908d-11e6-8f99-db4e9eff8038.png)

3. Sometimes scrolling down (too fast?) changes the line counts by one, no idea how this happens. As proof I took these two screenshots (taken only a minute apart from each other):
![screenshot from 2016-10-12 14 16 34](https://cloud.githubusercontent.com/assets/19438668/19311665/a06b7c38-908f-11e6-9c09-c28e8a3431b4.png)
![screenshot from 2016-10-12 14 17 41](https://cloud.githubusercontent.com/assets/19438668/19311676/a9c7efdc-908f-11e6-8048-aae7f835a69b.png)

I guess this is not a big problem, but interesting to notice...

stgrue commented:
1. Forcing a line break within a word would likely require a larger rework of how tokens are mapped to elements of the annotation page, so I am not sure how feasible this is. I suggest opening an issue but otherwise leaving it for now, since this situation is rather rare anyway and can easily be resolved by splitting the word manually.
2. This is likely caused by the insertion of whitespace that is not taken into account, I will look into it.
3. This is a known issue that is not caused by this change: https://github.com/annefried/swan/issues/302

Commit:
DA-289: Refactor and rework handling of line breaks

In order to fix several bugs and undesirable effects, this patch reworks large
parts of how line breaks are handled in the annotation view. Most importantly,
a new system for calculating the width of lines was added, so that the positions
of line breaks can now be determined with greater accuracy instead of relying on
coarse approximations of the average character width. Alongside the more general
refactoring, this has the following effects:
- The application does not crash anymore when the length of a word exceeds the
  length of the annotation box.
- Lines containing many characters of above-average length (e.g. in Chinese texts)
  are not cut off anymore.
- Line breaks are (in general) closer to the edge of the annotation box, resulting
  in a more economic use of horizontal space.
- The new system can be manually optimized for other languages and scripts as well.

stgrue commented:
Updated.
